File: src/electron/menu.ts
import { app, Menu, shell, BrowserWindow } from 'electron'
import type { MenuItemConstructorOptions } from 'electron'

/**
 * Custom macOS menu bar. Mac apps are expected to have one — and Cut /
 * Copy / Paste keyboard shortcuts come from the Edit menu's role-based
 * items, so we can't just remove the menu the way we do on Windows/Linux.
 *
 * Every Verko-specific item sends a string command to the focused window
 * over `app:menu-command`; the renderer listens once and dispatches to
 * its existing UI handlers. Accelerators are advisory — Electron consumes
 * them before they reach the renderer, so commands fire from menu shortcuts
 * AND from the renderer's own keydown listener identically.
 */
export function buildMacMenu(getWindow: () => BrowserWindow | null): Menu {
  const send = (cmd: string) => () => {
    getWindow()?.webContents.send('app:menu-command', cmd)
  }

  const template: MenuItemConstructorOptions[] = [
    {
      label: app.name,
      submenu: [
        { role: 'about' },
        { type: 'separator' },
        { label: 'Settings…', accelerator: 'Cmd+,', click: send('open-settings') },
        { type: 'separator' },
        { role: 'services' },
        { type: 'separator' },
        { role: 'hide' },
        { role: 'hideOthers' },
        { role: 'unhide' },
        { type: 'separator' },
        { role: 'quit' },
      ],
    },
    {
      label: 'File',
      submenu: [
        { label: 'New Conversation', accelerator: 'Cmd+N', click: send('new-conversation') },
        { type: 'separator' },
        { label: 'Find in Library', accelerator: 'Cmd+F', click: send('open-filter') },
        { label: 'Command Palette', accelerator: 'Cmd+K', click: send('open-command-palette') },
        { type: 'separator' },
        { role: 'close' },
      ],
    },
    {
      label: 'Edit',
      submenu: [
        { role: 'undo' },
        { role: 'redo' },
        { type: 'separator' },
        { role: 'cut' },
        { role: 'copy' },
        { role: 'paste' },
        { role: 'pasteAndMatchStyle' },
        { role: 'delete' },
        { role: 'selectAll' },
      ],
    },
    {
      label: 'View',
      submenu: [
        { label: 'Toggle Sidebar', accelerator: 'Cmd+\\', click: send('toggle-sidebar') },
        { label: 'Toggle Agent',   accelerator: 'Cmd+.',  click: send('toggle-agent') },
        { type: 'separator' },
        { role: 'togglefullscreen' },
        { type: 'separator' },
        { role: 'reload' },
        { role: 'toggleDevTools' },
      ],
    },
    {
      label: 'Window',
      submenu: [
        { role: 'minimize' },
        { role: 'zoom' },
        { type: 'separator' },
        { role: 'front' },
      ],
    },
    {
      label: 'Help',
      submenu: [
        {
          label: 'Documentation',
          click: () => { void shell.openExternal('https://github.com/CatVinci-Studio/Verko') },
        },
        {
          label: 'Report an Issue',
          click: () => { void shell.openExternal('https://github.com/CatVinci-Studio/Verko/issues/new') },
        },
      ],
    },
  ]

  return Menu.buildFromTemplate(template)
}

File: src-tauri/src/menu.rs
// Native macOS menu. Without one, WebKit refuses to map ⌘C / ⌘V / ⌘A to
// edit roles in text fields and Cmd+Q does nothing. Windows / Linux are
// frameless via `decorations: false`, so they get no menu — the in-app
// titlebar (TitleBar.tsx) is the entire chrome.
//
// Items are intentionally minimal: system roles only, no custom Verko
// commands. The renderer already owns its own keyboard-shortcut layer
// (Cmd+K / Cmd+F / Cmd+, etc) via window-level keydown listeners; a menu
// would just duplicate them.

#[cfg(target_os = "macos")]
use tauri::menu::{AboutMetadataBuilder, MenuBuilder, PredefinedMenuItem, SubmenuBuilder};
#[cfg(target_os = "macos")]
use tauri::AppHandle;

#[cfg(target_os = "macos")]
pub fn install(app: &AppHandle) -> tauri::Result<()> {
    let pkg = app.package_info();
    let about_metadata = AboutMetadataBuilder::new()
        .name(Some(pkg.name.clone()))
        .version(Some(pkg.version.to_string()))
        .copyright(Some("Copyright © CatVinci-Studio".into()))
        .build();

    let app_menu = SubmenuBuilder::new(app, &pkg.name)
        .item(&PredefinedMenuItem::about(app, None, Some(about_metadata))?)
        .separator()
        .item(&PredefinedMenuItem::services(app, None)?)
        .separator()
        .item(&PredefinedMenuItem::hide(app, None)?)
        .item(&PredefinedMenuItem::hide_others(app, None)?)
        .item(&PredefinedMenuItem::show_all(app, None)?)
        .separator()
        .item(&PredefinedMenuItem::quit(app, None)?)
        .build()?;

    let edit_menu = SubmenuBuilder::new(app, "Edit")
        .item(&PredefinedMenuItem::undo(app, None)?)
        .item(&PredefinedMenuItem::redo(app, None)?)
        .separator()
        .item(&PredefinedMenuItem::cut(app, None)?)
        .item(&PredefinedMenuItem::copy(app, None)?)
        .item(&PredefinedMenuItem::paste(app, None)?)
        .item(&PredefinedMenuItem::select_all(app, None)?)
        .build()?;

    let view_menu = SubmenuBuilder::new(app, "View")
        .item(&PredefinedMenuItem::fullscreen(app, None)?)
        .build()?;

    let window_menu = SubmenuBuilder::new(app, "Window")
        .item(&PredefinedMenuItem::minimize(app, None)?)
        .item(&PredefinedMenuItem::maximize(app, None)?)
        .separator()
        .item(&PredefinedMenuItem::close_window(app, None)?)
        .build()?;

    let menu = MenuBuilder::new(app)
        .item(&app_menu)
        .item(&edit_menu)
        .item(&view_menu)
        .item(&window_menu)
        .build()?;

    app.set_menu(menu)?;
    Ok(())
}

#[cfg(not(target_os = "macos"))]
pub fn install(_app: &tauri::AppHandle) -> tauri::Result<()> {
    Ok(())
}
